### PR TITLE
Gemfile resolution in Action.shell_out_should_use_bundle_exec (#7925)

### LINKED ISSUE #7925: I think Fastlane::Action.shell_out_should_use_bundle_exec is bugged.
Hi there,

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines]

I think there's a problem with `Fastlane::Action.shell_out_should_use_bundle_exec` and that it's not behaving the way it should.

In a test setup I have now, I'm installing fastlane in a bundle. Here's the (simplified) structure of the project:

```
+ Gemfile # has fastlane and a few other things
+ Dir/
\---+ fastlane
    \---+ Fastfile etc
```

The problem is that, when running `cd Dir && bundle exec fastlane`, `shell_out_should_use_bundle_exec` will try to find a Gemfile in the current directory (`Dir`) only, fail, and prevent all actions from executing through bundler. It's currently implemented this way:

```rb
File.exist?('Gemfile') && !Helper.contained_fastlane?
```

I believe it should be changed to something like this (loosely adapted from `Bundler::SharedHelpers::find_gemfile`):

```rb
if !Helper.contained_fastlane?
  given = ENV["BUNDLE_GEMFILE"] # If present, fastlane (or a parent process) was most likely invoked using bundle exec
  return true if given && !given.empty?

  previous = nil
  current  = File.expand_path(Pathname.pwd).untaint
  until !File.directory?(current) || current == previous
    filename = File.join(current, "Gemfile")
    return true if File.file?(filename)

    previous = current
    current = File.expand_path("..", current)
  end
end
return false
```

Am I wrong? If not, should I send a PR?

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                 | Value                                           |
| ------------------- | ----------------------------------------------- |
| OS                  | 10.12.2                                         |
| Ruby                | 2.3.1                                           |
| Bundler?            | true                                            |
| Git                 | git version 2.11.0                              |
| Installation Source | ~/.rbenv/versions/2.3.1/bin/fastlane            |
| Host                | Mac OS X 10.12.2 (16C67)                        |
| Ruby Lib Dir        | ~/.rbenv/versions/2.3.1/lib                     |
| OpenSSL Version     | OpenSSL 1.0.2h  3 May 2016                      |
| Is contained        | false                                           |
| Xcode Path          | /Applications/Xcode_8.2.app/Contents/Developer/ |
| Xcode Version       | 8.2                                             |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             | 🚫 |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
opt_out_usage
default_platform :ios

platform :ios do
  desc 'Run all the tests'
  lane :tests do |options|
    cocoapods
    scan
  end
end
# vim:syntax=ruby:et:sts=2:sw=2:ts=2:ff=unix:
```
</details>

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.9.0   | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| bundler                   | 1.13.7       |
| io-console                | 0.4.6        |
| rake                      | 12.0.0       |
| CFPropertyList            | 2.3.4        |
| ZenTest                   | 4.11.1       |
| RubyInline                | 3.12.4       |
| i18n                      | 0.7.0        |
| json                      | 1.8.6        |
| minitest                  | 5.10.1       |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| activesupport             | 4.2.7.1      |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.0        |
| babosa                    | 1.0.2        |
| claide                    | 1.0.1        |
| clamp                     | 0.6.5        |
| fuzzy_match               | 2.0.4        |
| nap                       | 1.1.0        |
| cocoapods-core            | 1.1.1        |
| cocoapods-deintegrate     | 1.0.1        |
| cocoapods-downloader      | 1.1.3        |
| cocoapods-plugins         | 1.0.0        |
| cocoapods-search          | 1.0.0        |
| cocoapods-stats           | 1.0.0        |
| netrc                     | 0.7.8        |
| cocoapods-trunk           | 1.1.2        |
| cocoapods-try             | 1.1.0        |
| colored                   | 1.2          |
| escape                    | 0.0.4        |
| fourflusher               | 2.0.1        |
| gh_inspector              | 1.0.3        |
| molinillo                 | 0.5.5        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.2        |
| cocoapods                 | 1.1.1        |
| dotenv                    | 2.1.2        |
| osx_keychain              | 1.0.1        |
| cocoapods-keys            | 2.0.0        |
| highline                  | 1.7.8        |
| commander                 | 4.4.3        |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| excon                     | 0.54.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.11.0       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.11.0       |
| fastimage                 | 2.0.1        |
| jwt                       | 1.5.6        |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.1.0        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| hurley                    | 0.2          |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| google-api-client         | 0.9.20       |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.2.0        |
| rubyzip                   | 1.2.0        |
| security                  | 0.1.3        |
| slack-notifier            | 1.5.1        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| word_wrap                 | 1.0.0        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| xcpretty-travis-formatter | 0.0.4        |
| mini_portile2             | 2.1.0        |
| nokogiri                  | 1.6.8.1      |
| slather                   | 2.3.0        |
| xcode-install             | 2.1.0        |
</details>

*generated on:* **2017-01-20**
</details>

### PR DESCRIPTION
**Pull request by jcayzac**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Description
This fixes #7925 by reimplementing `Action.shell_out_should_use_bundle_exec` in a way that closely matches how Bundler resolves the location of the Gemfile.

### Motivation and Context
Fixes #7925. Now, invoking e.g. `cocoapods(use_bundle_exec: true)` will properly invoke `bundle exec pod…` even if the Gemfile is found in an ancestor of the current directory or if fastlane itself was invoked using `bundle exec fastlane`.

### CODE REVIEW

**Comment by jcayzac** [comment]
Note: I had to reset my Xcode preferences for the tests to pass. The test `Scan Scan::DetectValues fetches the path from the Xcode config` makes a pretty strong assumption on the normal location of derived data and I believe this is a bug (of the test itself).

My Xcode preferences usually set the derived data to lie in the `build` folder relative to the current project, and the *Advanced* location settings request non-unique, custom build paths (`intermediates` and `products`). With those enabled, fastlane's tests fail:

```
Scan Scan::DetectValues fetches the path from the Xcode config
     Failure/Error: expect(derived_data).to match(%r{/.*Xcode/DerivedData/app-\w*$})

       expected "/Users/jcayzac/github/fastlane/fastlane/scan/examples/standard/build" to match /\/.*Xcode\/DerivedData\/app-\w*$/
       Diff:
       @@ -1,2 +1,2 @@
       -/\/.*Xcode\/DerivedData\/app-\w*$/
       +"/Users/jcayzac/github/fastlane/fastlane/scan/examples/standard/build"

     # ./scan/spec/detect_values_spec.rb:11:in `block (3 levels) in <top (required)>'
```

**Comment by hjanuschka** [comment]
@jcayzac nice catch! thx for the PR. could you please add a unit test for it to prevent regression's?
somwhere in `fastlane/spec/action_spec.rb`

**Comment by jcayzac** [comment]
@hjanuschka probably not until next week, and it will be a learning experience (I'm quite new to Ruby actually) :)

**Comment by hjanuschka** [comment]
@jcayzac no worries, it should be something like this.


```ruby
describe "shell_out_should_use_bundle_exec?" do
      it "should return false on contained version" do
        expect(FastlaneCore::Helper).to receive(:contained_fastlane?).and_return(true)
        expect(Fastlane::Action.shell_out_should_use_bundle_exec?).to eq(false)
      end
      it "should return true if no GEMFILE set" do
        with_env_values('BUNDLE_GEMFILE' => "someGemfile") do
          expect(Fastlane::Action.shell_out_should_use_bundle_exec?).to eq(true)
        end
      end
      it "should return true if Gemfile can be found" do
        with_env_values('BUNDLE_GEMFILE' => nil) do
          expect(File).to receive(:file?).and_return(true)
          expect(Fastlane::Action.shell_out_should_use_bundle_exec?).to eq(true)
        end
      end
      it "should return false if not contained, BUNDLE_GEMFILE not set and Gemfile could not be found" do
        with_env_values('BUNDLE_GEMFILE' => nil) do
          allow(File).to receive(:file?).and_return(false)
          expect(Fastlane::Action.shell_out_should_use_bundle_exec?).to eq(false)
        end
      end
    end
```

File: fastlane/lib/fastlane/action.rb [hunk 1]
```diff
@@ -118,9 +118,24 @@ def self.method_missing(method_sym, *arguments, &_block)
...
     def self.shell_out_should_use_bundle_exec?
-      return File.exist?('Gemfile') && !Helper.contained_fastlane?
+      unless Helper.contained_fastlane?
+        given = ENV["BUNDLE_GEMFILE"]
+        return true if given && !given.empty?
+
+        previous = nil
+        current  = File.expand_path(Pathname.pwd).untaint
+        until !File.directory?(current) || current == previous
+          filename = File.join(current, "Gemfile")
```
Comment (segiddins) [inline fastlane/lib/fastlane/action.rb]: Note that `gems.rb` is also a valid Gemfile name

**Review by endocrimes — CHANGES_REQUESTED** [review]
This should be good to go when also supporting `gems.rb`

File: fastlane/lib/fastlane/action.rb (hunk 1 above)
Comment (KrauseFx) [inline fastlane/lib/fastlane/action.rb]: Great point, thanks @segiddins, @jcayzac there is a method in `bundler` that allows you to easily get the name for the Gemfile in the current directory, let me know if you need any help 👍

**Comment by jcayzac** [comment]
@DanToml @segiddins @KrauseFx @hjanuschka I'm terribly sorry for the long delay. I think I got all the comments covered for now. `gems.rb` is handled and there are unit tests. Let me know if anything else is needed.

**Comment by jcayzac** [comment]
Actually I don't quite understand the raison d'être of `Action.shell_out_should_use_bundle_exec`. Why do you prevent actions from calling `bundle exec` when their options ask them to? It goes to the shell anyway.

If that's to lock users in with the versions of some gems that are in the contained version, then perhaps a better approach would be to not shell the commands out but call the gems programmatically instead, from the current ruby environment?

I'm not really happy with my pull request replicating Bundler's file resolution, when Bundler would do it just fine on its own if the `bundle exec` shell command was not being vetoed.

**Comment by jcayzac** [comment]
@segiddins @DanToml is there anything else required to land this in `master`?

**Comment by KrauseFx** [comment]
> This fixes #7925 by reimplementing Action.shell_out_should_use_bundle_exec in a way that closely matches how Bundler resolves the location of the Gemfile.

Is there any way we can **use** Bundler's implementation? `bundler` is already a dependency of _fastlane_, so we could call its methods directly.

**Comment by segiddins** [comment]
`Bundler.default_gemfile`?

**Comment by KrauseFx** [comment]
Thanks @segiddins 👍 

Hey @jcayzac, it looks like this is a great patch you have, but it should be updated to use the method @segiddins mentioned. Do you have the time to update the PR to make use of it? 

Thanks for all your contribution 🚀

**Comment by jcayzac** [comment]
Hi @KrauseFx!
Sorry for the long delay in getting back. Unfortunately I'm working on Android projects right now which are consuming all my time. I think @segiddins's suggestion makes total sense and should be trivial to implement but I can't commit to do that any time soon. Hopefully when things get less crazy at work I'll have some time 😬
I'd also love more information regarding my question in [this comment](https://github.com/fastlane/fastlane/pull/7926#issuecomment-279942147), if you don't mind. I can't help thinking something might be wrong and `shell_out_should_use_bundle_exec` should not exist to begin with.

**Comment by jcayzac** [comment]
Actually, now this project is CLA-encumbered and I have no intent on signing that, I'll close the PR.

**Comment by taquitos** [comment]
@jcayzac I'd love to understand your thoughts on the CLA. If there is something that needs to change, or should be made better, we can suggest updates to the lawyers. I see you're really busy right now, but if you ever get a chance to give us feedback, I'd be happy to follow up. You can reach me in our slack I'm @taquitos there too.

**Comment by jcayzac** [comment]
Hi @taquitos,
I have nothing against this particular CLA (I haven't read it), but I am not willing to sign a legally-binding agreement or any sort to simply notify an open-source project's authors that they can exercise their rights to take my changes to the project's source code, as the license already allows them. CLA are a cancer.

**Pull request closed without merging**